typescript/cli/src/rebalancer/executor/Executor.ts

```diff
@@ -0,0 +1,212 @@
+        throw new Error(`Signer ${signerAddress} is not a rebalancer`);
+      }
+
+      if (
+        (await originHypAdapter.getAllowedDestination(domain)) !== recipient
+      ) {
+        throw new Error(
+          `Destination ${recipient} for domain ${domain} is not allowed`,
+        );
+      }
+
+      if (!(await originHypAdapter.isBridgeAllowed(domain, bridge))) {
+        throw new Error(`Bridge ${bridge} for domain ${domain} is not allowed`);
+      }
+
+      // Skip this rebalance route if the amount is below the configured minimum threshold.
+      // This prevents dust amounts or economically unviable transfers
+      if (bridgeMinAcceptedAmount > amount) {
+        log(
+          `Route ${fromChain} → ${toChain} skipped: amount ${amount} below minimum threshold ${bridgeMinAcceptedAmount}`,
+        );
+
+        continue;
+      }
+
+      log(
+        `Getting rebalance quotes: domain=${domain}, amount=${amount}, bridge=${bridge}`,
+      );
+
+      let quotes: InterchainGasQuote[];
+
+      try {
+        quotes = await originHypAdapter.getRebalanceQuotes(
+          bridge,
+          domain,
+          recipient,
+          amount,
+          bridgeIsWarp,
+        );
+      } catch (error) {
+        throw new Error(
+          `Could not get rebalance quotes: ${(error as Error).message}`,
+        );
+      }
+
+      log(
+        `Populating rebalance transaction: domain=${domain}, amount=${amount}, bridge=${bridge}`,
+      );
+
+      const populatedTx = await originHypAdapter.populateRebalanceTx(
+        domain,
+        amount,
+        bridge,
+        quotes,
+      );
+
+      transactions.push({ signer, populatedTx });
+    }
+
+    // Early return if no valid routes were found to rebalance.
+    // This happens when all potential routes were skipped (e.g., due to minimum amounts)
+    if (transactions.length === 0) {
+      log('Rebalance skipped: No routes to execute');
+
+      return;
+    }
+
+    log('Estimating gas for all transactions');
+
+    // Estimate gas before sending transactions.
+    // This is mainly to check that the transaction will not fail before sending them.
+    const estimateGasResults = await Promise.allSettled(
+      transactions.map(async ({ signer, populatedTx }, i) => {
+        try {
+          await signer.estimateGas(populatedTx);
+          log(`Gas estimation succeeded for route ${i}`);
+        } catch (error) {
+          log(`❌ Could not estimate gas for route`, routes[i]);
+          throw error;
+        }
+      }),
+    );
+
+    if (estimateGasResults.some((result) => result.status === 'rejected')) {
+      throw new Error('❌ Could not estimate gas for some routes');
+    }
+
+    log('Sending transactions');
+    const results = await Promise.allSettled(
+      transactions.map(async ({ signer, populatedTx }, i) => {
+        log(`Sending transaction for route ${i}`);
+        try {
+          const tx = await signer.sendTransaction(populatedTx);
```

Comment:
can we use the MultiProvider instead of the signer? it includes this boilerplate as `handleTx()` and is included in every CLI command as `context.multiprovider`